### PR TITLE
linux audioshare: add granular selection, more options, better ui

### LINKED ISSUE #218: add support for blacklisting apps for audio screenshare
usually, when i'm in a discord call, it's with my partner, and both of us deafen each other since we speak over mumble and use discord's screensharing feature to watch youtube videos or play games. 

in order to prevent having to close the screenshare and re-open a new one whenever either of us decide to do something else, we usually use the "entire system" option to capture all audio all of the time, and then i have to open helvum (pipewire patchbay) to disconnect mumble from vesktop, so that my partner can't hear himself.

i noticed that vesktop automatically excludes itself from its sound being streamed over screenshare, presumably to prevent echo from the voice chat. if vesktop has this ability, maybe there can be an option to automatically exclude *any* app the user wants (in my case, mumble)? it could maybe be a text box where you can enter the names of binaries to exclude, or even maybe a dropdown list that lists all of the current running apps and lets you pick which ones to opt out of sharing.

### LINKED ISSUE #619: [Linux][X11] Audio Sinks as Stream Input
### Discord Account

apostle_of_floof

### Motivation

When selecting an audio source for a screen share, you can only select "Entire System" or individual apps.
Entire Audio obviously isn't exactly the greatest choice when using sinks, as it even connects the sinks discord is piped into (meaning viewers will hear themselves in the stream).
Entire System

The second option of selecting individual apps isn't working as great either. Some apps/games just don't appear in the list (E.g. Rabbit and Steel), and some other games take ALL nodes related to the game, including input/microphone nodes.
Helldivers

### Solution

Enable selecting audio sinks and/or indidivual devices as audio source for streams, similar to how it works in discords voice settings.

### Alternatives

I tried manually setting the connections in qpwgraph, but starting a new stream/changing the audio source will reset all connections, forcing me to manually set/remove each connection again. Exclusive Mode fixes this issue, but forces you to connect new nodes manually to devices and such.

### Additional context

_No response_

### Request Agreement

- [X] I have searched the existing issues and found no similar issue
- [X] This is not a plugin request
- [X] This is not a Vencord feature request

### PR DESCRIPTION
**Pull request by Curve**
- [x] Bump Venmic
- [x] Granular Selection
- [x] Allow to select multiple audio sources
- [x] Additional Exclusion Options
- [x] Redesign Screen-Share Picker
- [x] New "Ignore Virtual" Setting
- [x] Add missing venmic settings
- [x] Make AudioShare settings persistent 

---

image
image
image
image

---

> [!NOTE]
> The settings in the screenshot do not represent the defaults :)

Fixes #619
Fixes #218

### CODE REVIEW

**Comment by Curve** [comment]
> i think it would also be useful to update the Pulseaudio server warning to add a recommendation to switch to pipewire via https://gist.github.com/the-spyke/2de98b22ff4f978ebf0650c90e82027e#install

Will do!

File: src/main/venmic.ts [hunk 1]
```diff
@@ -69,47 +71,59 @@ function getRendererAudioServicePid() {
...
+        ignore_devices: settings.audioIgnoreDevices
     };
 
-    if (workaround) {
-        data.workaround = [
-            { key: "application.process.id", value: pid },
-            { key: "media.name", value: "RecordStream" }
-        ];
+    if (settings.audioIgnoreInputMedia ?? true) {
+        data.exclude?.push({ "media.class": "Stream/Input/Audio" });
```
Comment (Vendicated) [inline src/main/venmic.ts:92]:
```suggestion
        data.exclude.push({ "media.class": "Stream/Input/Audio" });
```

File: src/main/venmic.ts [hunk 2]
```diff
@@ -69,47 +71,59 @@ function getRendererAudioServicePid() {
...
+    }
+
+    if (settings.audioIgnoreVirtual) {
+        data.exclude?.push({ "node.virtual": "true" });
```
Comment (Vendicated) [inline src/main/venmic.ts:96]:
```suggestion
        data.exclude.push({ "node.virtual": "true" });
```

File: src/main/venmic.ts [hunk 3]
```diff
@@ -69,47 +71,59 @@ function getRendererAudioServicePid() {
...
+        data.workaround = [{ "application.process.id": pid, "media.name": "RecordStream" }];
     }
 
     return obtainVenmic()?.link(data);
 });
 
-ipcMain.handle(IpcEvents.VIRT_MIC_START_SYSTEM, (_, workaround?: boolean, onlyDefaultSpeakers?: boolean) => {
+ipcMain.handle(IpcEvents.VIRT_MIC_START_SYSTEM, (_, exclude: Node[]) => {
     const pid = getRendererAudioServicePid();
+    const settings = Settings.store;
```
Comment (Vendicated) [inline src/main/venmic.ts:108]: why not use destructuring?
```suggestion
    const { audioOnlySpeakers, audioIgnoreDevices, ... } = Settings.store;
```

File: src/shared/settings.d.ts [hunk 4]
```diff
@@ -30,6 +30,16 @@ export interface Settings {
...
+
+    audioWorkaround?: boolean;
+    audioGranularSelect?: boolean;
+
+    audioIgnoreVirtual?: boolean;
+    audioIgnoreDevices?: boolean;
+    audioIgnoreInputMedia?: boolean;
+
+    audioOnlySpeakers?: boolean;
+    audioOnlyDefaultSpeakers?: boolean;
```
Comment (Vendicated) [inline src/shared/settings.d.ts:42]:
```suggestion
    venmic: {
    	workaround?: boolean;
    	granularSelect?: boolean;
    	ignoreVirtual?: boolean;
    	ignoreDevices?: boolean;
    	ignoreInputMedia?: boolean;
    	onlySpeakers?: boolean;
    	onlyDefaultSpeakers?: boolean;
    }
```

**Pull request merged**